### PR TITLE
Scatter lowering

### PR DESCRIPTION
**Pull request by rmoyard**
**Description of the Change:**

In the quantum dialect, I add a lowering for MHLO scatter to the MLIR tensor dialect. The reference for the implementation can be found here https://github.com/openxla/stablehlo/blob/main/docs/spec.md#scatter

**Benefits:**

This unlocks Jax array index with at and apply functions on it, for example:

```
    @qjit
    def add_multiply(l: jax.core.ShapedArray((3,), dtype=float), idx: int):
        res = l.at[idx].multiply(3)
        res2 = l.at[idx].add(2)
        return res + res2

    res = add_multiply(jnp.array([0, 1, 2]), 2)
    >>> [0, 2, 10]
```

**Possible Drawbacks:**

Some edge cases not properly covered.

### CODE REVIEW

**Comment by dime10** [comment]
We might want to wait until #216 is merged, since that PR is fairly heavy and close to being merged. It shouldn't be too much work to adapt this PR to the new compiler driver.

File: .github/workflows/check-catalyst.yaml
Comment (rmoyard) [inline .github/workflows/check-catalyst.yaml:223]: I have to readd that.

**Comment by dime10** [comment]
[sc-41325]

File: .github/workflows/check-catalyst.yaml
Comment (erick-xanadu) [inline .github/workflows/check-catalyst.yaml:266]: I think changes here are redundant. Are we only changing the order of the "get cached mhlo" and "get cached llvm"?

File: mlir/test/Catalyst/ScatterTest.mlir
Comment (erick-xanadu) [inline mlir/test/Catalyst/ScatterTest.mlir:181]: Should we check that there are not more than 24 calls? Also, is this supposed to be inside of a for-loop?

File: frontend/test/pytest/test_scatter.py
Comment (erick-xanadu) [inline frontend/test/pytest/test_scatter.py:46]: Can you add a test for multiple dimensions? Matrix and maybe a 3dimensional array?

File: .github/workflows/check-catalyst.yaml
Comment (erick-xanadu) [reply .github/workflows/check-catalyst.yaml:266]: Isn't "Get Cached MHLO Build" duplicated now? (See lines below)

**Review by erick-xanadu — COMMENTED** [review]
Still looking at it. Some small comments first.

File: mlir/lib/Quantum/Transforms/ScatterPatterns.cpp
Comment (grwlf) [inline mlir/lib/Quantum/Transforms/ScatterPatterns.cpp:47]: Returning failure would probably lead to passing the ScatterOp down the pipeline. I imagine that users would see kind of 'unkown op' errors then. Would it be better to call `emitError` with a verbose message explaining the failure reason?

File: .github/workflows/check-catalyst.yaml
Comment (rmoyard) [reply .github/workflows/check-catalyst.yaml:266]: Removed!

File: frontend/test/pytest/test_scatter.py
Comment (rmoyard) [reply frontend/test/pytest/test_scatter.py:46]: Yes I am doing it

File: mlir/lib/Quantum/Transforms/ScatterPatterns.cpp
Comment (rmoyard) [reply mlir/lib/Quantum/Transforms/ScatterPatterns.cpp:47]: Yes I will add that

File: mlir/test/Catalyst/ScatterTest.mlir
Comment (rmoyard) [reply mlir/test/Catalyst/ScatterTest.mlir:181]: Yes thanks!

Comment (rmoyard) [reply mlir/test/Catalyst/ScatterTest.mlir]: I will replace the for loop with a scf for loop as part of the technical debt

**Review by erick-xanadu — APPROVED** [review]
Awesome! Great work @rmoyard!

**Pull request merged**
